### PR TITLE
testkit: add a WithTiKV flag to support run unit test on real TiKV

### LINKED ISSUE #35646: Can I run tidb unit test cases with a real TiKV?
## General Question

There are a lot of test cases in the tidb repo.
It's run with the unistore engine.
Can I run it with TiKV? and how to do that?

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #35646 

Problem Summary:

Add a `-with-tikv` flag to the unit test.

### What is changed and how it works?

After this change, I can run **any** test case with a real TiKV, for example:

```
cd executor
go test -run TestFix31530 -with-tikv 127.0.0.1:2379
``` 

I can verify out test cases with real tikv. Here is how I do that:

1. Create a docker image:

```
cat Dockerfile
FROM buildpack-deps:focal as base

COPY pd-server /bin/
COPY tikv-server /bin/
COPY start.sh /

ENV LANG C.UTF-8
ENV LC_ALL C.UTF-8

FROM base AS branch-version-1
```

```
cat start.sh
#!/bin/bash
set -eu

/bin/pd-server --client-urls http://0.0.0.0:2379 --data-dir "/data/pd" &
/bin/tikv-server --status-addr 0.0.0.0:20180  --pd 0.0.0.0:2379 -A 127.0.0.1:20160 -C "/conf/tikv.toml" -s "/data/tikv"
```

2. build the image

```
docker build -t tikv-img .
```

3. Run the tikv in container

```
docker run --net=host --rm --name tmp1 tikv-img:bootstrap /start.sh
```

`--net` use the local port, `--rm` means remove the container after the test is done.

4. Run unit test with the tikv docker container

For example:

```
go test -run TestFix31530 -with-tikv 127.0.0.1:2379
```

5. Automate the test script

I modify the `make ut` script to automate the tests.
Create the tikv docker container before test and cleanup them.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: testkit/mockstore.go [hunk 1]
```diff
@@ -17,20 +17,42 @@
...
 	"github.com/pingcap/tidb/session"
+	"github.com/pingcap/tidb/store/driver"
 	"github.com/pingcap/tidb/store/mockstore"
 	"github.com/stretchr/testify/require"
 	"github.com/tikv/client-go/v2/oracle"
 	"github.com/tikv/client-go/v2/tikv"
 )
 
+// WithTiKV flag makes the test case run with real TiKV
+var WithTiKV = flag.String("with-tikv", "", "whether tests run with real TiKV")
```
Comment (hawkingrei) [inline testkit/mockstore.go:35]: In fact, @tisonkun has moved the real-tikv test case on [here](https://github.com/pingcap/tidb/blob/master/tests/realtikvtest/testkit.go#L43). If you would like to run the real-tikv test case. all are under the same folder.

Comment (tiancaiamao) [reply testkit/mockstore.go:35]: The background is, I need to test the paging protocol with the real tikv.
So I need rewrite all the test case or move them all?
A more reasonable solution is just add a flag ... if you don't use it, it doesn't change anything.

Comment (tisonkun) [reply testkit/mockstore.go:35]: @tiancaiamao would you like to change CI logic? Or adding these code for conveniently running tests with real tikv locally?

Comment (tisonkun) [reply testkit/mockstore.go:35]: Also, `with-tikv` is a string flag while its help info said `"whether tests run with real TiKV"`, which is confusing...

Perhaps

```go
// This flag is only used for debugging locally with real tikv cluster.
var tikvAddr = flag.String("tikv-addr", "", "address of tikv cluster, if set, running test with real tikv cluster")
```

**Review by bb7133 — APPROVED** [review]
LGTM

File: testkit/mockstore.go (hunk 1 above)
Comment (tiancaiamao) [reply testkit/mockstore.go]: > Also, `with-tikv` is a string flag while its help info said `"whether tests run with real TiKV"`, which is confusing...
> 
> Perhaps
> 
> ```go
> // This flag is only used for debugging locally with real tikv cluster.
> var tikvAddr = flag.String("tikv-addr", "", "address of tikv cluster, if set, running test with real tikv cluster")
> ```

Updated.

Comment (tiancaiamao) [reply testkit/mockstore.go]: > @tiancaiamao would you like to change CI logic? Or adding these code for conveniently running tests with real tikv locally?

If you want to see the full changes, it's in my branch https://github.com/tiancaiamao/tidb/tree/test-with-tikv
But not all the test cases can run under TiKV, I check and document the result here:

https://pingcap.feishu.cn/docx/doxcnEzyRXsTdCJ2EE6EZwIP2oe

I'm not adding a new CI pipeline, I just make the code base more friendly for those who want to run some test cases with a real TiKV.

**Review by xhebox — APPROVED** [review]

**Comment by hawkingrei** [comment]
If the ```WithTikv``` is added in the testkit, We will run the with-tikv tests with the all tidb repo. so we have to build all the test cases to run them. It will take more time to build it. so We suggest the with tikv cases in the specific folders. so It can reduce the objects to be compiled.

**Comment by hawkingrei** [comment]
@tiancaiamao Please ```gofmt`` the ```tools/check/ut.go```. I find it is useless for ```ut.go``` when ```gofmt``` is run in the root folder.

**Pull request merged**